### PR TITLE
Inline RTCSessionDescriptionInit description

### LINKED ISSUE #33962: Broken link to RTCSessionDescriptionInit, RTCLocalSessionDescriptionInit
### MDN URL

https://developer.mozilla.org/en-US/docs/Web/API/RTCSessionDescription

### What specific section or headline is this issue about?

_No response_

### What information was incorrect, unhelpful, or incomplete?

The page for `RTCPeerConnection.createAnswer` has a broken link to `RTCSessionDescriptionInit` because there is no separate page for it

Actually there is a bigger issue here. MDN needs to clarify the relationship between `RTCSessionDescription`, `RTCSessionDescriptionInit`, and `RTCLocalSessionDescriptionInit`.

### What did you expect to see?

A page for `RTCSessionDescriptionInit`, also a separate page for `RTCLocalSessionDescriptionInit` would be nice. Also clarify the relationship between `RTCSessionDescription`, `RTCSessionDescriptionInit`, and `RTCLocalSessionDescriptionInit`.

### Do you have any supporting links, references, or citations?

https://w3c.github.io/webrtc-pc

### Do you have anything more you want to share?

Here is my understanding of this issue (since this is long, I will use a collapsible section):

<details>
<summary>Details</summary>

Relevant WebIDL definitions for RTCSessionDescription, RTCSessionDescriptionInit, and RTCLocalSessionDescriptionInit (from https://w3c.github.io/webrtc-pc):

```webidl
[Exposed=Window]
interface RTCSessionDescription {
  constructor(RTCSessionDescriptionInit descriptionInitDict);
  readonly attribute RTCSdpType type;
  readonly attribute DOMString sdp;
  [Default] RTCSessionDescriptionInit toJSON();
};

dictionary RTCSessionDescriptionInit {
  required RTCSdpType type;
  DOMString sdp = "";
};

dictionary RTCLocalSessionDescriptionInit {
  RTCSdpType type;
  DOMString sdp = "";
};

[Exposed=Window]
interface RTCPeerConnection : EventTarget  {
...
  Promise<RTCSessionDescriptionInit> createOffer(optional RTCOfferOptions options = {});
  Promise<RTCSessionDescriptionInit> createAnswer(optional RTCAnswerOptions options = {});

  Promise<undefined> setLocalDescription(optional RTCLocalSessionDescriptionInit description = {});
  readonly attribute RTCSessionDescription? localDescription;
  readonly attribute RTCSessionDescription? currentLocalDescription;
  readonly attribute RTCSessionDescription? pendingLocalDescription;
  Promise<undefined> setRemoteDescription(RTCSessionDescriptionInit description);
  readonly attribute RTCSessionDescription? remoteDescription;
  readonly attribute RTCSessionDescription? currentRemoteDescription;
  readonly attribute RTCSessionDescription? pendingRemoteDescription;
...
}
```

**Understanding RTCSessionDescription:**

- **Core Interface:** RTCSessionDescription is the fundamental interface representing a session description in WebRTC signaling.
- **Properties:** It has two essential properties:
  - `type`: Specifies the SDP type (e.g., "offer", "answer", "pranswer", or "rollback").
  - `sdp`: Contains the actual SDP string, which is meaningless for the "rollback" type.

**RTCSessionDescriptionInit and RTCLocalSessionDescriptionInit**

These names closely resemble RTCSessionDescription, but they serve a different purpose. Here's a breakdown:

* **Dictionaries, Not Classes:** Unlike RTCSessionDescription (an interface), these are dictionaries. You don't need a specific instance to create them; any conforming object will suffice.
* **Arguments for Setting Session Descriptions:** They are primarily used as arguments for the `RTCPeerConnection.setLocalDescription()` and `RTCPeerConnection.setRemoteDescription()` methods, allowing you to set session descriptions without creating a full-fledged RTCSessionDescription object.
* **Return Values for Creating SDP:** `RTCSessionDescriptionInit` is also the return type for `RTCPeerConnection.createOffer()` and `RTCPeerConnection.createAnswer()`, which generate SDP strings. These generated SDPs are typically used for either setting a local description or sending them to a remote peer.

**Why Separate Interfaces?**

The separation between `RTCSessionDescriptionInit` and `RTCLocalSessionDescriptionInit` might seem unnecessary and confusing at first. Here's the reasoning:

* **Evolution of WebRTC:** WebRTC has been around for over a decade and is constantly evolving. Initially, `RTCPeerConnection.setLocalDescription()` always required both an SDP string and a type.
* **Parameterless setLocalDescription:** A few years ago, this method was updated to allow omitting any arguments. This feature, sometimes called "parameterless setLocalDescription," is part of the "Perfect Negotiation" functionality. It enables users to skip the separate calls to `RTCPeerConnection.createOffer()` and `RTCPeerConnection.setLocalDescription()`. They can simply use the parameterless version of `setLocalDescription()`.
* **RTCLocalSessionDescriptionInit for Flexibility:** To accommodate this change, `RTCLocalSessionDescriptionInit` was introduced. Unlike `RTCSessionDescriptionInit`, it doesn't even require a `type` property. In WebIDL, if you call `RTCPeerConnection.setLocalDescription()` without any arguments, it's equivalent to calling `RTCPeerConnection.setLocalDescription({})`. This can be seen as a way to enhance WebIDL's flexibility.

**In Summary:**

* **RTCSessionDescription:** Core interface representing a session description.
* **RTCSessionDescriptionInit:** Dictionary used to create session descriptions and as a return type for `createOffer()` and `createAnswer()`.
* **RTCLocalSessionDescriptionInit:** Dictionary specifically used for setting local session descriptions, allowing omission of the `type` property.

</details>

### MDN metadata

<details>
<summary>Page report details</summary>

* Folder: `en-us/web/api/rtcsessiondescription`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/API/RTCSessionDescription
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/api/rtcsessiondescription/index.md
* Last commit: https://github.com/mdn/content/commit/acfe8c9f1f4145f77653a2bc64a9744b001358dc
* Document last modified: 2023-07-07T07:19:19.000Z

### PR DESCRIPTION
**Pull request by bc-lee**
### Description

Add missing explanation of RTCSessionDescriptionInit and RTCLocalSessionDescriptionInit dictionaries. Also correct several links referencing those interfaces.
### Related issues and pull requests
Fixes #33962

### CODE REVIEW

**Review by skyclouds2001 — COMMENTED** [review]
I think the two dictionaries are better not to placed in separate pages, as they both only contains two properties and both only used in one method

**Review by Josh-Cena — CHANGES_REQUESTED** [review]
Hi! Unfortunately we don't document dictionaries on MDN, unless they are reused by a lot of methods (and even in such case we would usually just copy the property list around). Could you inline the description in the places where they are used instead?

**Review by Josh-Cena — COMMENTED** [review]
We never mention the dictionary's name at all, because it's not a code entity. It should look like this:

```
  - : An object with the following properties:
    - `property1`
      - : What it does
```

And then for the errors, you would just say "the `sessionDescription` parameter is invalid".

**Comment by bc-lee** [comment]
Do you mean I should update the PR to remove any references to the directory name `RTCSessionDescriptionInit`? It existed before I submitted my PR.

**Comment by Josh-Cena** [comment]
Yes, they should be removed.

**Comment by bc-lee** [comment]
I'll keep my PR open for now, but I'm not sure if I can make any more changes, especially considering that the distinction of dictionary names is required since they are part of the WebRTC API. I believe it is important to retain the dictionary name in the documentation. If someone wants to take over the PR, I'm happy to close it.

**Comment by Josh-Cena** [comment]
Please see past changes where dictionaries were removed, for example https://github.com/mdn/content/pull/30080, https://github.com/mdn/content/pull/30079

**Comment by hamishwillee** [comment]
@bc-lee FYI @Josh-Cena is correct - the fact that the spec mentions these dictionaries is irrelevant. Specs are written for browser developers, while MDN interprets the specification for webapp developers.

In most cases we've found that having separate dictionaries is unhelpful for web app developers - they have to jump around to find relevant information, and often the compatibility data for the dictionary depends on where it is used. Because of that, the best way to properly track compatibility is to expand the dictionary.

We're not dogmatic about it. For some APIs it does actually make sense to define the objects passed to APIs as dictionaries. Generally this is when you have a general API that can take numerous possible objects depending on the context in which it is used.
This is not the case for WebRTC.

We won't be merging this as is. I'm sure it will be useful though if modified to match MDN conventions.

**Comment by bc-lee** [comment]
Two days ago, a pull request relevant to this one (#34051) was merged. I'd like to hear from @Josh-Cena and @hamishwillee on whether you believe the merged PR achieves the same level of consistency as requested in this pull request.

**Comment by Josh-Cena** [comment]
@bc-lee, do you have time to come back to this?

**Comment by bc-lee** [comment]
Yes, but not today or tomorrow (atm I'm busy with other things). I'll come back to this in a few days. 

https://github.com/mdn/content/pull/33955#issuecomment-2284985438

File: files/en-us/web/api/rtcsessiondescription/rtcsessiondescription/index.md [hunk 1]
```diff
@@ -45,7 +45,7 @@ This example uses the constructor to convert an SDP offer into an
 `RTCSessionDescription` object.
 
 > **Note:** This is no longer necessary, however;
-> {{domxref("RTCPeerConnection.setLocalDescription()")}} and other methods which take
+> [`RTCSessionDescriptionInit`](/en-US/docs/Web/API/RTCSessionDescription#rtcsessiondescriptioninit) and other methods which take
```
Comment (Josh-Cena) [inline files/en-us/web/api/rtcsessiondescription/rtcsessiondescription/index.md:48]:
```suggestion
> {{domxref("RTCPeerConnection.setLocalDescription()")}} and other methods which take
```

**Review by Josh-Cena — APPROVED** [review]
I've simply inlined the descriptions in each page and reordered some sections. I did not do other major changes. Thanks for the PR!

**Pull request merged**